src/Plugin/facets/query_type/HierarchyQueryType.php

```diff
@@ -0,0 +1,215 @@
+      $plugin_definition,
+      $container->get('entity_field.manager'),
+      $container->get('entity_type.manager'),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * This is mostly copied from SearchApiString however there are a few
+   * differences. We check if the operator is with or without hierarchy
+   * and change the operator used in the query to use all values returned
+   * in the hierarchy of the defined value.
+   *
+   * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+   */
+  public function execute() {
+    $query = $this->query;
+
+    // Only alter the query when there's an actual query object to alter.
+    if (empty($query)) {
+      return;
+    }
+    $original_operator = $this->facet->getQueryOperator();
+    $field_identifier = $this->facet->getFieldIdentifier();
+    $exclude = $this->facet->getExclude();
+
+    if ($query->getProcessingLevel() === QueryInterface::PROCESSING_FULL) {
+      // Set the options for the actual query.
+      $options = &$query->getOptions();
+      $options['search_api_facets'][$field_identifier] = $this->getFacetOptions();
+    }
+
+    // Add the filter to the query if there are active values.
+    $active_items = $this->facet->getActiveItems();
+
+    if (count($active_items)) {
+      // Convert the operator to use in case the equivalent operator
+      // with hierarchy is used.
+      if ($original_operator == HiearchicalMultiselectWidget::AND_WITH_HIERARCHY_OPERATOR) {
+        $operator = ListPresetFilter::AND_OPERATOR;
+      }
+      elseif ($original_operator == HiearchicalMultiselectWidget::OR_WITH_HIERARCHY_OPERATOR) {
+        $operator = ListPresetFilter::OR_OPERATOR;
+      }
+      elseif ($original_operator == HiearchicalMultiselectWidget::NONE_WITH_HIERARCHY_OPERATOR) {
+        $operator = ListPresetFilter::NOT_OPERATOR;
+      }
+      else {
+        $operator = $original_operator;
+      }
+
+      $filter = $query->createConditionGroup($operator, ['facet:' . $field_identifier]);
+      foreach ($active_items as $value) {
+        if (str_starts_with($value, '!(')) {
+          /** @var \Drupal\facets\UrlProcessor\UrlProcessorInterface $urlProcessor */
+          $url_processor = $this->facet->getProcessors()['url_processor_handler']->getProcessor();
+          foreach (explode($url_processor->getDelimiter(), substr($value, 2, -1)) as $missing_value) {
+            // Note that $exclude needs to be inverted for "missing".
+            // Maintain original logic but alter the operator to use in case
+            // we detect the operator is with hierarchy.
+            if ($operator == $original_operator) {
+              $filter->addCondition($this->facet->getFieldIdentifier(), $missing_value, !$exclude ? '<>' : '=');
+            }
+            else {
+              $hierarchy_values = $this->getHierarchy($missing_value, $this->facet);
+              $filter->addCondition($this->facet->getFieldIdentifier(), $hierarchy_values, !$exclude ? 'NOT IN' : 'IN');
```

Comment:
Here (and below where you use the hierarchy values), you should check if there are any. So `getHierarchy()` can return an empty array. What do you want to do to the query in that case? THink on this for each operator because it might not make sense always to send an empty array as condition